server/src/controllers/aplus.ts

```diff
@@ -0,0 +1,217 @@
+// SPDX-FileCopyrightText: 2024 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+import {Request, Response} from 'express';
+import {z, ZodError} from 'zod';
+import {TypedRequestBody} from 'zod-express-middleware';
+
+import {
+  AplusExerciseData,
+  AplusGradeSourceType,
+  HttpCode,
+  IdSchema,
+  NewAplusGradeSourceArraySchema,
+  NewAplusGradeSourceData,
+  NewGrade,
+} from '@/common/types';
+import {fetchFromAplus, validateAplusCourseId} from './utils/aplus';
+import {validateAttainmentPath} from './utils/attainment';
+import {APLUS_API_URL} from '../configs/environment';
+import AplusGradeSource from '../database/models/aplusGradeSource';
+import {ApiError} from '../types';
+
+/**
+ * Responds with AplusExerciseData
+ *
+ * @throws ApiError(400|502)
+ */
+export const fetchAplusExerciseData = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  const aplusCourseId = validateAplusCourseId(req.params.aplusCourseId);
+
+  const exercisesRes = await fetchFromAplus<{
+    results: {
+      id: number;
+      display_name: string;
+      exercises: {
+        difficulty: string;
+      }[];
+    }[];
+  }>(`${APLUS_API_URL}/courses/${aplusCourseId}/exercises?format=json`);
+
+  // There doesn't appear to be a better way to get difficulties
+  const difficulties = new Set<string>();
+  for (const result of exercisesRes.data.results) {
+    for (const exercise of result.exercises) {
+      if (exercise.difficulty) {
+        difficulties.add(exercise.difficulty);
+      }
+    }
+  }
+
+  const exerciseData: AplusExerciseData = {
+    modules: exercisesRes.data.results.map(result => ({
+      id: result.id,
+      name: result.display_name,
+    })),
+    difficulties: Array.from(difficulties),
+  };
+
+  res.json(exerciseData);
+};
+
+/** @throws ApiError(400|404|409) */
+export const addAplusGradeSources = async (
+  req: TypedRequestBody<typeof NewAplusGradeSourceArraySchema>,
+  res: Response
+): Promise<void> => {
+  const newGradeSources: NewAplusGradeSourceData[] = req.body;
+  for (const newGradeSource of newGradeSources) {
+    await validateAttainmentPath(
+      req.params.courseId,
+      String(newGradeSource.attainmentId)
+    );
+  }
+
+  await AplusGradeSource.bulkCreate(newGradeSources);
+
+  res.sendStatus(HttpCode.Created);
+};
+
+/**
+ * Responds with NewGrade[]
+ *
+ * @throws ApiError(400|404|409|422|502)
+ */
+export const fetchAplusGrades = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  let attainmentIds: number[] = [];
+  try {
+    attainmentIds = z
+      .array(IdSchema)
+      .parse(JSON.parse(String(req.query.attainments)));
+  } catch (e) {
+    if (e instanceof Error) {
+      throw new ApiError(e.message, HttpCode.BadRequest);
+    }
+    if (e instanceof ZodError) {
+      throw new ApiError(
+        e.issues.map(issue => issue.message),
+        HttpCode.BadRequest
+      );
+    }
```

Comment:
This is never reached because of the above `(e instanceof Error)`.
I would also recommend using zod `safeparse` or `safeparseAsync` to get rid of one try catch.

Reply:
The problem here is `JSON.parse`, which will throw an error if `req.query.attainments` is invalid. As far as I know, there is no easy safe way to parse JSON (I'm sure there are libraries though), so we might as well use the same try/catch block for Zod errors as well.

Good catch about this being unreachable. Do we even need to handle the Zod error separately? `ZodError` has a `message` field anyway, which returns `issues` as a JSON string.